### PR TITLE
[Python] use get_buffer to fetch buffer when the buffer is None

### PR DESCRIPTION
**Pull request by AnandInguva**
Fixes: https://github.com/apache/beam/issues/25315
Fixes: https://github.com/apache/beam/issues/23228

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: sdks/python/apache_beam/runners/portability/fn_api_runner/fn_runner.py [hunk 1]
```diff
@@ -852,6 +852,9 @@ def _execute_bundle(self,
               (consuming_stage_name, consuming_transform, buffer_id))
         # We enqueue all of the pending output buffers to be scheduled at the
         # MAX_TIMESTAMP for the downstream stage.
+        # buffer is None. Empty PCollection. Bug
```
Comment (AnandInguva) [inline sdks/python/apache_beam/runners/portability/fn_api_runner/fn_runner.py:855]:
```suggestion
```

File: sdks/python/apache_beam/runners/portability/fn_api_runner/fn_runner.py [hunk 2]
```diff
@@ -852,6 +859,9 @@ def _execute_bundle(self,
...
+        if not buffer:
```
Comment (AnandInguva) [inline sdks/python/apache_beam/runners/portability/fn_api_runner/fn_runner.py:863]:
```suggestion
```

File: sdks/python/apache_beam/runners/portability/fn_api_runner/fn_runner.py [hunk 3]
```diff
@@ -852,6 +859,9 @@ def _execute_bundle(self,
...
+          buffer = ListBuffer(None)
```
Comment (AnandInguva) [inline sdks/python/apache_beam/runners/portability/fn_api_runner/fn_runner.py:864]:
```suggestion
```

**Comment by AnandInguva** [comment]
R: @pabloem cc: @tvalentyn

**Comment by AnandInguva** [comment]
I tested the TFX test mentioned at https://github.com/apache/beam/issues/25315#issuecomment-1514524080. It is working

File: sdks/python/apache_beam/runners/portability/fn_api_runner/fn_runner.py [hunk 4]
```diff
@@ -840,6 +841,12 @@ def _execute_bundle(self,
           # so we create a copy of the buffer for every new stage.
           runner_execution_context.pcoll_buffers[buffer_id] = buffer.copy()
           buffer = runner_execution_context.pcoll_buffers[buffer_id]
+        # When the buffer is not in the pcoll_buffers, it means that the
+        # it could be an empty PCollection. In this case, get the buffer using
+        # the buffer id and transform id
+        if buffer is None:
+          buffer = bundle_context_manager.get_buffer(buffer_id, transform_id)
+          assert buffer is not None
```
Comment (liferoad) [inline sdks/python/apache_beam/runners/portability/fn_api_runner/fn_runner.py:849]: can we log this with a clear error message?

Comment (AnandInguva) [reply sdks/python/apache_beam/runners/portability/fn_api_runner/fn_runner.py]: I was meant to remove the assertion from there. It was just there for debugging purpose.

Comment (liferoad) [reply sdks/python/apache_beam/runners/portability/fn_api_runner/fn_runner.py]: I see. Thanks!

**Comment by pabloem** [comment]
this LGTM! Thanks so much @AnandInguva - can you add a test case? Feel free to merge after adding the test case.

**Review by liferoad — APPROVED** [review]

**Comment by AnandInguva** [comment]
Lint error is not related to this PR

**Pull request merged**